### ISSUE TITLE
Is there a way to detect long press of volume button?

### ISSUE THREAD

**Issue by hyd00** [jpsim/JPSVolumeButtonHandler#27]

**Comment by jpsim** [comment]
I haven't used this library myself for a few years now, so I'm not sure how you could go about doing this. I hope someone else who's used it more actively recently can chime in. Maybe @joestelmach?

**Comment by joestelmach** [comment]
I think you'll continue to get notifications when you hold the button down, so you can do something like this:

```
-(void)onVolumeUp: {
  if(self.timer) {
    [self.timer invalidate];
  }
  self.timer = [NSTimer scheduledTimerWithTimeInterval:0.5
                                               target:self
                                             selector:@selector(onTimerFire)
                                             userInfo:nil
                                              repeats:NO];
}

-(void)onTimerFire {
  NSLog(@"Long Press");
  self.timer = nil;
}
```

Keep in mind that the onVolumeUp method will need to run on the same thread every time for the invalidate call to work (https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSTimer_Class/)

**Comment by hyd00** [comment]
Thanks man! This works!

**Comment by hyd00** [comment]
@joestelmach Please excuse me for asking, Your solution works perfectly, but it also fires the function on short press. How can we differentiate between long and short press? Is there a way?

**Comment by hyd00** [comment]
I did this but it's not working properly:

`//Volume button handler
    self.volumeButtonHandler = [JPSVolumeButtonHandler volumeButtonHandlerWithUpBlock:^{
        // Volume Up Button Pressed
        ++secondsElapsed;
        [self onVolumeUp];
    } downBlock:^{
        // Volume Down Button Pressed
    }];

-(void)onVolumeUp {
    if(self.volButtonTimer) {
        [self.volButtonTimer invalidate];
    }
    self.volButtonTimer = [NSTimer scheduledTimerWithTimeInterval:0.5
                                                  target:self
                                                selector:@selector(onTimerFire)
                                                userInfo:nil
                                                 repeats:NO];
    
}

-(void)onTimerFire {
    NSLog(@"Long Press %d",secondsElapsed);
    if (secondsElapsed > 1)
    {
        //Do Something
    }
    else secondsElapsed = 0;
    
}
`

The DO SOMETHING block fires only when I release the button after long press. What happening is, I believe, that `OnTimeFire` gets called very fast, even before the secondElapsed increments to 2.

Any guides how can I fix this? @joestelmach 

Thanks in anticipation :)

**Comment by joestelmach** [comment]
you just need to incremement a count within the ```if(self.volButtonTimer)```, then check that the count is more than one within ```onTimerFire```

**Comment by hyd00** [comment]
Tried. Same issue. When release the button after long press, it fires the desired method. :(

**Comment by hyd00** [comment]
Still waiting for any more help :(

**Comment by joestelmach** [comment]
paste your code and let's look at it

**Comment by hyd00** [comment]
`//Volume button handler

self.volumeButtonHandler = [JPSVolumeButtonHandler volumeButtonHandlerWithUpBlock:^{
        // Volume Up Button Pressed
        [self onVolumeUp];
    } downBlock:^{
        // Volume Down Button Pressed
    }];

-(void)onVolumeUp {
    if(self.volButtonTimer) {
        secondsElapsed = 1; //To detect long press
        [self.volButtonTimer invalidate];
    }
    self.volButtonTimer = [NSTimer scheduledTimerWithTimeInterval:0.5
                                                  target:self
                                                selector:@selector(onTimerFire)
                                                userInfo:nil
                                                 repeats:NO];

}

-(void)onTimerFire {
    NSLog(@"Long Press %d",secondsElapsed);
    if (secondsElapsed > 1)
    {
        //Do Something
    }
    else secondsElapsed = 0;

}`

**Comment by joestelmach** [comment]
How will secondsElapsed ever be greater than 1?

**Comment by hyd00** [comment]
Sorry.Edited.Please,check

**Comment by timov7** [comment]
Hi, maybe this is a bit late, but maybe someone might find useful. If you would like to have methods called when there is long press detected you might verify secondsElapsed in onVolumeUp method like this: 
// let's say when count is 3 is enough to admit that it was long press
`if (self.volButtonTimer) {
                if (self.secondsElapsed == 3) {
                    self.secondsElapsed = 0;
                    // call your method here
                }
                self.secondsElapsed ++;
                [self.volButtonTimer invalidate];
            }
            
            self.volButtonTimer = [NSTimer scheduledTimerWithTimeInterval:0.8 target:self selector:@selector(onTimerFire) userInfo:nil repeats:NO];

- (void)onTimerFire {
    self.volButtonTimer = nil;
   if (self.timeElapsed != 0) {
        self.timeElapsed = 0;
    }
    } else {
       // do what you want on simple press
    }
}`

**Comment by hyd00** [comment]
@timov7 let me trh and get back. Did you try it urself though?

**Comment by timov7** [comment]
Yes, it is working for me.

**Comment by hyd00** [comment]
It looks good enough to detect long press yes. But does it also help u detect short press?

**Comment by timov7** [comment]
You can detect short press in onTimeFire method
`if (self.secondsElapsed != 0) {
        self.secondsElapsed = 0;
    } else {
        NSLog(@"short");
        // short press method
    }
`

**Comment by timov7** [comment]
Yes, you need to make timer nil each time onTimeFire gets called. Otherwise secondsElapsed will increase continuously in onVolumeUp method.

**Comment by hyd00** [comment]
Still confused lol ... But thanks a lot for pitching in :)

P.S [You're Beautiful - James Blunt](https://www.youtube.com/watch?v=oofSnsGkops)

**Comment by timov7** [comment]
I am happy to help. :) See my edited answer for onTimeFire method.
